drivers/timer/mchp_xec_rtos_timer.c

```diff
@@ -0,0 +1,271 @@
+/*
+ * Copyright (c) 2018 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#include <drivers/system_timer.h>
+#include <sys_clock.h>
+#include <spinlock.h>
+#include <arch/arm/cortex_m/cmsis.h>
+
+#define DEBUG_RTOS_TIMER
+
+#define COUNTER_MAX 0x0fffffffUL
+#define TIMER_STOPPED 0xf0000000UL
+
+#if 0
+#define CYC_PER_TICK (CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC	\
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+#else
+
+#define CYC_PER_TICK_M ((CONFIG_SYS_CLOCK_HW_CYCLES_PER_SEC * 256)	\
+		      / CONFIG_SYS_CLOCK_TICKS_PER_SEC)
+
+#if (CYC_PER_TICK_M & 0xFF) > 0x7F
+#define CYC_PER_TICK ((CYC_PER_TICK_M >> 8) + 1)
+#else
+#define CYC_PER_TICK (CYC_PER_TICK_M >> 8)
+#endif
+
+#endif
+
+#define MAX_TICKS ((COUNTER_MAX / CYC_PER_TICK) - 1)
+#define MAX_CYCLES (MAX_TICKS * CYC_PER_TICK)
+
+/* Minimum cycles in the future to try to program.  Note that this is
+ * NOT simply "enough cycles to get the counter read and reprogrammed
+ * reliably" -- it becomes the minimum value of the LOAD register, and
+ * thus reflects how much time we can reliably see expire between
+ * calls to elapsed() to read the COUNTFLAG bit.  So it needs to be
+ * set to be larger than the maximum time the interrupt might be
+ * masked.  Choosing a fraction of a tick is probably a good enough
+ * default, with an absolute minimum of 1k cyc.
+ */
+#define MIN_DELAY MAX(32, (CYC_PER_TICK/16))
+
+#define TICKLESS (IS_ENABLED(CONFIG_TICKLESS_KERNEL) &&			\
+		  !IS_ENABLED(CONFIG_QEMU_TICKLESS_WORKAROUND))
+
+/* VAL value above which we assume that a subsequent COUNTFLAG
+ * overflow seen in CTRL is real and not an artifact of wraparound
+ * timing.
+ */
+#define VAL_ABOUT_TO_WRAP 1
+
+static struct k_spinlock lock;
+
+static volatile u32_t last_load;
+
+static volatile u32_t cycle_count;
+
+static volatile u32_t announced_cycles;
+
+static volatile u32_t elapsed_counts;
+
+static ALWAYS_INLINE void timer_clear_intr(void)
+{
+	GIRQ23_REGS->SRC = MCHP_RTMR_GIRQ_VAL;
+	GIRQ23_REGS->RESULT = GIRQ23_REGS->SRC;
+	NVIC_ClearPendingIRQ(RTMR_IRQn);
```

Comment:
Why hit the NVIC?  That bit should clear automatically once the CPU interrupt has been delivered and the source device is deasserted, right?  In general ARM drivers don't need to do this.